app/src/main/java/com/example/erabook/AuthenticationViewModel.kt

```diff
@@ -1,22 +1,78 @@
 package com.example.erabook
 
+import android.content.Context
+import androidx.fragment.app.Fragment
 import androidx.lifecycle.LiveData
 import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
+import com.bumptech.glide.Glide.init
+import com.google.android.gms.auth.api.signin.GoogleSignIn
+import com.google.android.gms.auth.api.signin.GoogleSignInClient
+import com.google.android.gms.auth.api.signin.GoogleSignInOptions
+import com.google.firebase.FirebaseException
 import com.google.firebase.auth.FirebaseAuth
 import com.google.firebase.auth.FirebaseUser
+import com.google.firebase.auth.GoogleAuthProvider
 
+private const val RC_SIGN_IN = 9001
 class AuthenticationViewModel : ViewModel() {
     private val firebaseAuth = FirebaseAuth.getInstance()
-    val userLiveData: LiveData<FirebaseUser?> = MutableLiveData()
+    private val _userLiveData = MutableLiveData<FirebaseUser?>()
+    val userLiveData: LiveData<FirebaseUser?> = _userLiveData
+    private val _isTaskSuccessful = MutableLiveData<Boolean>()
+    val isTaskSuccessful:LiveData<Boolean?> = _isTaskSuccessful
+    private val _exception = MutableLiveData<String>()
+    val exception: LiveData<String> = _exception
+    private lateinit var mGoogleSignInClient: GoogleSignInClient
 
     init {
         firebaseAuth.addAuthStateListener { firebaseAuth ->
-            (userLiveData as MutableLiveData).value = firebaseAuth.currentUser
-
+            _userLiveData.value = firebaseAuth.currentUser
         }
     }
 
+    fun signInWithEmailAndPassword(email: String, password:String){
+        firebaseAuth.signInWithEmailAndPassword(email,password)
+            .addOnCompleteListener { task ->
+                if (task.isSuccessful){
+                    _isTaskSuccessful.value = true
+                }else{
+                    _isTaskSuccessful.value = false
+                    _exception.value = task.exception?.message
```

Comment:
What if exception is null? On that case user will now know what is happening since you haven't provided any default message. Please provide a default message.

Reply:
the exception value is set only if there is an error or when the task is not successful. When the task is not successful so the user is not logged in gets the error message and sets that value to the _exception variable. I don't think the _exception value when it's accessed it's null because the only case where we are using it is where the task is not successful therefore the value of the _exception gets updated before accessing it.

we don't have a default in this case because we have to know what the message contains so we can show the user what is going on in the fragment(check code lines in the LoginFragment  76-93)